### ISSUE TITLE
Bert 정리

### ISSUE THREAD

**Issue by ComputerMath** [ComputerMath/Math-repository#2]
<Dataset & Dataloader>

main.py에서 argparser로 들어온 min_rating = 0 / min_uc = 5 / min_sc = 0 / split = 'leave_one_out' / 을 dataset의 생성자의 인자를 통해 설정한다.

해당 dataset에 있어서의 전처리는 dataloader_factory에서 dataset_factory함수를 통해서 dataset 생성자를 만들어낸 뒤, dataloader에서 train, validation, test dataloader을 각각 만들어낼 때 ML1MDataset 클래스의 load_dataset 함수에서 이루어진다. 

전처리는 다음과 같다.
1. load_ratings_df

'http://files.grouplens.org/datasets/movielens/ml-1m.zip' 에 있는 ratings.dat 파일을 통해 data를 추출한다. 각 데이터의 한 단위는 '::'를 단위로 user id (uid), movie id (sid), rating, timestamp 가 분리되어있다. 이에 대한 분리는 pandas 라이브러리를 이용하여 datasets/ml_1m.py 에의 load_ratings_df 함수에서 작동한다.

2. making_implicit

3. filter_triplet

4.

**Comment by ComputerMath** [comment]
negative sampling

dataloader의 생성자에서, negative sample을 만드는 과정이 진행된다.
![image](https://user-images.githubusercontent.com/29303833/70332595-512f6900-1885-11ea-9bd3-5ea9da698676.png)

negative sample은 쉽게 이야기하자면, 어떤 user에 대해서, 해당 유저가 보지 않은 영화 목록들 중 커맨드의 argument로 들어온 sample_size (여기서는 100)만큼 중복없이 리스트를 만들어낸다.

해당 데이터는 이후 train시 metric관련해서 이용되므로 추후 설명할 것이다.
Dataloader은 train / validation / test 로 분리되며, train은 BertTrainDataset 클래스를 이용하며 validation과 test는 BertEvalDataset 클래스를 이용한다.

여기서 핵심은 BertTrainDataset이 train 시 mask처리를 위한 전처리를 __getitem__에서 하는 것이다.
![image](https://user-images.githubusercontent.com/29303833/70333387-b899e880-1886-11ea-8c06-afc221f7b4a3.png)

main.py에서의 argument parser로 온 mask_prob는 0.15이다
이를 기준으로 해당 for문 내부를 풀어서 설명해보자면, Dataloader에서 Dataset의 user에 해당하는 Movie sequence 를 추출해낼 때에는
0.12의 확률로 token에는 <mask>와 label에는 <Sequence의 Original Movie id>
0.015의 확률로 token에는 <Random Movie id> 와 label에는 <Sequence의 Original Movie id>
0.015의 확률로 token에는 <Original Movie id>와 label에는 <Sequence의 Original Movie id>
나머지 0.85의 확률에는 token에는 <Original Movie id>와 label에는 0 을 넣는다.

이후 getitem에서, argparser로 준 max_len (여기서는 100) 만큼 길이를 설정하기 위해, sequence 길이가 100 미만 시에는 앞부분에 padding(0)을 채워주고, 100 초과시에는 마지막 100개의 sequence로 채운채로 token과 label을 반환한다.

한편 BertEvalDataset에서는 __getitem__ 시 user에 해당하는 index에 대하여, user의 movie list를 반환하며 (반환1), validation과 test를 위한 movie id 1개와 negative sampling (여기서는 100개)를 하나의 list로 반환하며 (반환2), 첫번째의 원소 1 1개와 100개의 0을 하나의 list로 이어 반환한다 (반환3)
해당 반환값 3가지는 trainer에서 metric을 구하는데 이용된다. (이후 설명)


negative sampler을 이용하여 추후 trainer 모듈에서 metric을 구하는데 이용한다.

**Comment by ComputerMath** [comment]
![image](https://user-images.githubusercontent.com/29303833/70354618-64f1c400-18b3-11ea-8c60-7b9704ececeb.png)

1. Embedding
들어오는 input에 대해 우선 embedding layer을 거친다.
여기서 embedding은 (해당 모델에서는) TokenEmbedding과 (vocab size x embed_size), PositionalEmbedding(max_len x embed_size)  두 embed 를 거친 과정을 거친 값을 더한 뒤, dropout 을 취한다 (여기서는 0.1)

2. Transformer
Bert 모델은 기본적으로 트랜스포머로 이루어져있다. 
여기서는 transformer block을 4개만큼 쌓았으며,
일반적인 트랜스포머의 모델에 대해서 encoder 부분의 구조를 따왔다.

이 구조는 MultiheadAttention과, PositionwiseFeedForward로 크게 구성된다.


Muiti head attention모델의 구조는 들어오는 input에 대해 위 그림과 같은 attention 구조를 갖게 되며, 
batch normalization와 같은 형태로 skip connection으로 함게 이루어져 있다. (sublayer)

*이 때 batch normalization에 있어서 파라미터들은 학습 가능하다. (평균과 표준편차를 이용하기는 하지만, 표준화 후의 곱셈과 덧셈은 학습하는 parameter이다)

![image](https://user-images.githubusercontent.com/29303833/70356650-34605900-18b8-11ea-9fc2-7de54d03f92f.png)

**Comment by ComputerMath** [comment]
![image](https://user-images.githubusercontent.com/29303833/70357344-fd8b4280-18b9-11ea-91d4-3848365fb7ec.png)

위의 설명은 다음과 같은 그림으로 요약가능하다

**Comment by ComputerMath** [comment]
앞에서 정의한 모델로부터 distribution으로부터, CrossEntropyLoss 함수를 통해 train에 있어서의 loss를 구한다. 

![image](https://user-images.githubusercontent.com/29303833/70368235-6dfe8780-18eb-11ea-9bc8-9b8870c405ac.png)

validation과 test를 위한 data는 마지막 두 data에 대해서 각각하므로 마지막 time unit에 대한 값을 기준으로 모델에 대한 평가를 한다.
평가 기준에 대한 것은 recall과 ndcg (normalized discounted cumulative gain) 를 이용한다
